felix/bpf-gpl/conntrack_cleanup.c

```diff
@@ -0,0 +1,276 @@
+			} else {
+				CALI_DEBUG("  EXPIRED: Reverse NAT entry, queued for deletion. (Forward NAT entry not yet seen.)");
+			}
+		}
+		break;
+	}
+
+	return 0;
+}
+
+// process_ccq_entry processes an entry in the "cleanup queue" map.  The map
+// is keyed on the reverse NAT entry's key, with the value storing the forward
+// entry's key (or a zero value if the forward entry is missing).
+static long process_ccq_entry(void *map, const void *key, void *value, void *ctx)
+{
+	// Map stores mapping from reverse key to forward key (if known).
+	const struct calico_ct_key *rev_key = key;
+	const struct calico_ct_key *fwd_key = value;
+	struct ct_iter_ctx *ictx = ctx;
+
+	// It might be a few ms since we queued this entry, recheck it to make sure
+	// it is still expired.
+	const struct calico_ct_value *rev_value = cali_ct_lookup_elem(rev_key);
+	if (rev_value) {
+		__u64 age = sub_age(ictx->now, rev_value->last_seen);
+		__u64 age_s = age/1000000000ull;
+#ifdef IPVER6
+		CALI_DEBUG("Re-checking: proto=%d [%pI6]:%d", rev_key->protocol, &rev_key->addr_a, rev_key->port_a);
+		CALI_DEBUG("  <->[%pI6]:%d age=%ds", &rev_key->addr_b, rev_key->port_b, age_s);
+#else
+		CALI_DEBUG("Re-checking: proto=%d %pI4:%d", rev_key->protocol, &rev_key->addr_a, rev_key->port_a);
+		CALI_DEBUG("  <->%pI4:%d age=%ds", &rev_key->addr_b, rev_key->port_b, age_s);
+#endif
+		__u64 max_age = calculate_max_age(rev_key, rev_value, ictx);
+		__u64 max_age_s = max_age/1000000000ull;
+		if (age < max_age) {
+			// Race with a packet, CT entry now live again.
+			CALI_DEBUG("  RESURRECTED: entry no longer expired (max_age=%d).", max_age_s);
+			goto out;
+		}
+		CALI_DEBUG("  EXPIRED: cleaning up forward/reverse entries (max_age=%d).", max_age_s);
+	} else {
+		CALI_DEBUG("  MISSING: lookup failed.");
+	}
+
+	// Still expired, delete both entries.  The forward key might be a dummy
+	// all-zeros key but we know that key won't exist so we just let
+	// cali_ct_delete_elem handle that.
+	if (cali_ct_delete_elem(fwd_key) == 0) {
+		ictx->num_kvs_deleted_nat_fwd++;
+	}
+	if (cali_ct_delete_elem(rev_key) == 0) {
+		ictx->num_kvs_deleted_nat_rev++;
+	}
+
+out:
+	// Always remove the entry in our scratch map.
+	cali_ccq_delete_elem(key);
+	return 0;
+}
+
+// conntrack_cleanup is a BPF program that cleans up expired conntrack entries.
+// It does a single pass of the conntrack map, checking each entry for expiry.
+// Normal entries are deleted immediately if they are expired.  NAT entries are
+// queued for deletion in the cali_ccq map (so that forward and reverse entries
+// can be cleaned up together).
+__attribute__((section("tc"))) int conntrack_cleanup(struct __sk_buff *skb)
+{
+	struct ct_iter_ctx ictx = {};
+	bpf_skb_load_bytes(skb, 0, &ictx, sizeof(ictx));
+	if (ictx.now == 0) {
+		// Caller didn't provide a fixed time (as used in tests), use current
+		// time.
+		ictx.now = bpf_ktime_get_ns();
+	}
```

Comment:
If it is used in UTs only, perhaps this could be more expressive? But you would need to list it in `bpf-gpl/list-ut-objs` and compile a test version. So whatever is your preference.

```suggestion
#ifndef UNITTEST
		ictx.now = bpf_ktime_get_ns();
	}
#endif
```

Reply:
Decided not to build a UT binary.  Just seemed like overkill for one minor tweak (especially when the conntrack cleaner is not on critical packet path).